Implement video-length filter using ffprobe

Skaronator opened the pull request:
This implements two new filter options to filter videos based on their actual length before downloading them.

This fixes #4928

One interesting issue I noticed is that Motion JPEG exist. This is a video format that often only contains a single frame. Hence, I implemented the minimum frame logic.

shelvacu commented:
I see a few problems with this approach. The request sent by ffprobe will not have any of the headers (most importantly cookies), the retry logic is missing, and this will 'double up' every request.

The proper (much more annoying to implement) way to do this would be to include ffmpeg as a library, and then do the requests in python and pass the data to the ffmpeg library.

I think it might make sense to include this nearly as-is (it still does something useful), but I think it should be marked with warnings in the documentation (maybe even rename the options something like `experimental-min-length`) and log a warning if `headers["Cookies"]` is set

Skaronator commented:
Thanks for the feedback. I'll look into it

Skaronator commented:
@shelvacu do you have an example link/page that explicitly requires specific headers? Thanks

shelvacu commented:
Every site that requires a login will have to send a `Cookie` header. Other than that I'm not sure. I would suspect some sites insist on a `Referer` header but I'm not sure how to check which extractors add that.

Skaronator commented:
I've implemented the same retry mechanism as the standard HTTP downloader. The code has been moved to a separate file for better readability, and I've also added the header logic.

Please feel free to test it out, and any feedback would be greatly appreciated, @shelvacu  @mikf

shelvacu reviewed (approved):
Looks good, I hope this gets merged.

Skaronator commented:
Fixed an issue while handling an error. Otherwise the code so far is stable in my longer tests.

shelvacu commented on gallery_dl/ffprobe.py:
```diff
@@ -57,8 +57,7 @@ def get_video_length(obj, url):
         except subprocess.CalledProcessError as e:
-            msg = "ffprobe failed: " + e
-            print(e)
+            msg = "ffprobe failed: " + str(e)
```
you removed the `print` as well

Skaronator replied:
This was intentional since logging is already done in the loop in line 44 using the logger object.

Skaronator commented:
Rebased

Skaronator commented:
Rebased on master (v1.30.3) and adjusted for latest codebase changes. Tested successfully with coomer - video length filtering works as expected.

mikf committed:
[weibo] add 'text' option (#8422)

mikf committed:
[dt] replace 'parse_compat' with compat workaround

mikf committed:
[dt] update representation of invalid datetime objects

mikf committed:
[dt] use 'parse_datetime_iso()' for ISO formats

mikf committed:
[tests/formatter] update results

mikf committed:
merge branch 'dt': move datetime utils into separate module

- use 'datetime.fromisoformat()' when possible (#7671)
- return a datetime-compatible object for invalid datetimes
  (instead of a 'str' value)

mikf committed:
replace 'util.re()' with 'text.re()'

remove unnecessary 'util' imports

mikf committed:
use f-strings when building 'pattern'

mikf committed:
improve arguments for relative '__import__()'

using None for 'fromlist' is slightly faster than an empty tuple

featherbutt committed:
[pp:metadata] add 'newline' option (#8439)

* Add configuration setting to control newline characters in metadata.
* update docs/configuration
* introduce 'open()' method
* add 'newline' test

mikf committed:
[e621] add 'artist' & 'artist-search' extractors (#8448)

mikf committed:
[wikimedia] add 'format=original' to file URLs (#5512)

add 'format' option

mikf committed:
[wikimedia] implement config lookup for fandom/wikigg sites (#7283)

{
    "extractor": {
        "fandom": {
            "filename": "..."
        }
    }
}

mikf committed:
[wikimedia] fix 'limit' config lookup

rework config handling in general

mikf committed:
[wikimedia] provide 'lang' metadata (#7283)

mikf committed:
[wikimedia] restrict default 'format' to fandom/wikigg (#5512)

https://github.com/mikf/gallery-dl/commit/eefd8f9c0038f5aa1dee1f2cfc7f27dbe0a0f261#commitcomment-168635597

PrivateAccount85642650 committed:
[realbooru] fix 'tags' for video posts (#8455)

* fix realbooru tags for video posts
* fix lines too long
* combine 'tags' and 'tags_<category>' extraction
    - extract tag categories independent of 'tags' option
* add 'video' & 'animated GIF' tests

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

Nedrozak committed:
[instagram] extract 'subscription' metadata for story/highlight items (#8459)

mikf committed:
[cyberfile] add 'shared' extractor (#8323)

https://github.com/mikf/gallery-dl/issues/8323#issuecomment-3366857916

mikf committed:
[cyberfile:folder] support subfolders (#8323)

https://github.com/mikf/gallery-dl/issues/8323#issuecomment-3366857916

add 'recursive' option

mikf committed:
[fanbox:creator] add 'offset' option (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3456132723

mikf committed:
[shimmie2] support 'soybooru.com' (#8467)

mikf committed:
[common] fix 'AttributeError'

fixes regression introduced in 98d3354575048d5288beb27b078f15a28b55ed6d

mikf committed:
[paheal] fix 'AttributeError'

fixes regression introduced in 98d3354575048d5288beb27b078f15a28b55ed6d

mikf committed:
[imgbox] match direct links (#8474)

mikf committed:
[imagebam] fix 'filename' & 'extension' for names without ext (#8476)

mikf committed:
[imagevenue] improve error for dead images (#8477)

mikf committed:
[imagehosts] add and use 'not_found()' method

mikf committed:
[itaku] add 'order' option

mikf committed:
[patreon] add 'order-posts' option (#7856 #8482)

mikf committed:
[path] improve 'exists()' performance

call 'os.stat()' directly

mikf committed:
[path] treat broken symlinks as existing files (#8490)

mikf committed:
[text] reject long filename extensions (#8491)

fixes regression introduced in 3252ead7c74d257a45e8e15a30d884a3a528c2d1
ref bc868e7bb8ad82d9d2bef7a609aa2deb50fca647

moonshiner committed:
[cyberdrop] update domain to 'cyberdrop.cr' (#8496)

* added cr to list of cyberdrop TLDs
* update
- example URLs
- tests
- supportedsites

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[pp:metadata] document '"mode": "print"'

4052068d5a9ca0e3e20d2f26b62a31b64b4e5e7d

mikf committed:
[formatter] add 'Xb' format specifier - 'X' for bytes (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3479362186

mikf committed:
[formatter] add 'Lb' format specifier - 'L' for bytes

mikf committed:
[postimg] force 'postimg.cc' domain (#8505)

mikf committed:
[imagehost] improve 'filename' & 'extension' handling

mikf committed:
[silverpic] force '.net' TLD & fix extraction

mikf committed:
[imhentai:search] support '/advanced-search/' URLs (#8507)

mikf committed:
[postimg] fix extraction (#8505)

mikf committed:
[cien] fix 'creator' & 'recent' extractors (#8524)

update '_pagination_articles()' end element

mikf committed:
[deviantart:gallery] match URLs with query parameters (#8514)

gengenson-code committed:
Added support for sex.com's feed (cookie required) (#8519)

* Added Extractor for sex.com feed
* Removed old comment
* add test URL
* update supportedsites
* simplify & fix flake8 newlines
* warn about missing session cookie

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[misskey] remove base-class constructor

use 'self.groups[-1]' instead

mikf committed:
[jpgfish] update domain to 'jpg7.cr' (#8530)

mikf committed:
[common] allow general ISO 8601 values for 'date-min' & '-max'

and parse them as UTC times instead of local time

mikf committed:
[misskey] add 'text-posts' option (#8516)

do not request text-only notes by default

mikf committed:
[myhentaigallery] add 'tag' extractor (#8537)

mikf committed:
[misskey] add 'date-min' & 'date-max' options (#8516)

TODO: implement 'order-posts' option

dev-KingMaster committed:
[arena] add 'channel' extractor (#5847 #8509)

* feat: issue 5847 site support for are.na
* flake8 lint error fix
* class name error fix
* update
    - prevent unnecessary request to 'page_url'
    - fix pagination
    - simplify block extraction code
    TODO: - rewrite without GalleryExtractor
          - extractors for Blocks, Users, etc
* supportedsites
* tests
* rename to 'channel' extractor
* update site title to 'Are.na'
* prioritize attachments

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[arena] update metadata structure

- provide data in 'block', 'channel', 'user', etc dicts
- convert block dates to 'datetime' objects
- update directory/filename/archive formats

mikf committed:
[tests/results] improve '{…, …, ...}' tests

fix error when including 'dict' type as possible result

mikf committed:
[scripts/generate_result] don't reorder "small" dicts/lists

mikf committed:
[bellazon] remove query parameters from attachment IDs (#8544)

mikf committed:
[bellazon] improve 'filename' & 'extension' (#8544)

https://github.com/mikf/gallery-dl/issues/8544#issuecomment-3522617104

- use 'extension' from file URL (except attachments)
- strip everything after the first '.' from non-URL filenames

mikf committed:
[job] add 'keywords-global' option (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3505732704

mikf committed:
[imagehosts] provide 'post_url' metadata (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3520249958

mikf committed:
[pixhost] extract 'directory' metadata (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3520249958

mikf committed:
[rule34us:tag] support URLs with empty 'q' query parameter (#8546)

mikf committed:
[docs/configuration] update 'twitter.users' example (#8548)

Skaronator committed:
Merge remote-tracking branch 'upstream/master' into ffprobe

mikf committed:
[job] apply 'keywords-eval' to 'keywords-global' values (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3527246051

mikf committed:
[instagram] improve error for non-existent profiles (#8550)

mikf committed:
[reddit] fix "KeyError: 'media_metadata'" for embeds (#8551)

mikf committed:
[flickr] extract public API key from website (#7564 #7649 #7700 #8553)

this breaks 'oauth:flickr' with the default key,
but it allows downloading without custom key / Flickr Pro

mikf committed:
[instagram] fix "AttributeError: 'videos_dash'" (#8561)

fixes regression introduced in ab3e34e9c332ca3951f680f0b0378ec8c9509bd6

SpiffyChatterbox committed:
[cfake] add support (#707 #6021 #8549 #8430)

* Create cfake.py
* Update __init__.py
* Create cfake.py tests
* update
    - simplify & combine code
    - adjust 'pattern': use '[^/?#]', match lines and groups
    - generalize example URLs
    - update default filenames
* update docs/supportedsites
* update test results

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[reddit] use REST API by default (#8559)

mikf committed:
[fanbox:redirect] disable cookie usage (#8565)

mikf committed:
[xasiat] fix 'IndexError' for albums without category (#8569)

mikf committed:
[rawkuma] update extractors to new site layout (#8568)

vorsatile committed:
[aryion] add 'search' extractor (#8567)

* [aryion] Implement search extractor.
* [aryion] Update capabilities.
* [aryion] Adjust example.
* fix flake8 errors

* update & simplify
    - use existing '_pagination_next()'
    - remove '_pagination_search()'
    - update 'search[…]' metadata
* add tests

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[aryion:search] simplify further

- skip 'build_query()' step
- add underscores to prefixes

mikf committed:
[mangadex] add 'data-saver' option (#8573)

MyFinalBellyache committed:
[sxypix] add support (#4507 #8391 #8574)

* Add support for sxypix.com
* update & simplify
    - text.extract -> text.extr
    - self.request().json() -> self.request_json()
    - use self.groups
    - remove __init__
    - list comprehension
* update test results

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

MyFinalBellyache committed:
[okporn] add support (#8575)

* Add support for ok.porn
* update
    - adjust pattern/example/page_url
    - prevent unnecessary request
    - extract 'tags'
    - text.extract -> text.extr
    - return list comprehension result
* update supportedsites name
* update test results

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

MyFinalBellyache committed:
[pornstarstube] add support (#8576)

* Add support for pornstars.tube
* update
    - adjust pattern/example/page_url
    - extract '' & '' metadata
    - prevent unnecessary request
    - fix result URLs
    - return list comprehension result
* update supportedsites
* update test results
* fix extractor name

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[Danbooru] add 'media-asset' extractor (#8580)

mikf committed:
[fansly] handle posts without accountId/contentId/attachments (#8572)

MyFinalBellyache committed:
[scripts/init] Fix error running with default values (#8583)

Fixes:
```sh
python scripts/init.py newsite https://newsite.com
# [INFO] gallery_dl/extractor/newsite.py
# [ERROR] TypeError: 'NoneType' object is not iterable
# Traceback (most recent call last):
#   File "/gallery-dl/scripts/init.py", line 58, in init_extractor
#     fp.write(func(args))
#              ^^^^^^^^^^
#   File "/gallery-dl/scripts/init.py", line 88, in generate_module
#     {generate_extractors(args)}\
#      ^^^^^^^^^^^^^^^^^^^^^^^^^
#   File "/gallery-dl/scripts/init.py", line 108, in generate_extractors_basic
#     for subcat in args.subcategories:
# TypeError: 'NoneType' object is not iterable
# [INFO] gallery_dl/extractor/__init__.py
# [INFO] test/results/newsite.py
```

MyFinalBellyache committed:
[eporner] add support (#8581)

* [eporner] Add support
* Line-length <80 chars
* Fix flake8 indentation
* update
    - generalize example URL
    - strip "Photo Gallery" from 'title'
    - extract 'id'
* update supportedsites
* update test results

octylFractal committed:
[patreon] include full metadata with each URL (#4286 #8498)

* include full metadata object with each url
    Fixes #4286
* rename 'url_metadata' to 'file'

mikf committed:
[tiktok] ignore empty 'music' entries (#8571)

mikf committed:
[simpcity] fix starting from a specific page (#8599)

with the default '"order-posts": "desc"'

mikf committed:
[belazon] fix starting from a specific page

with the default '"order-posts": "desc"'

mikf committed:
[simpcity] fix 'content' for first post of a thread

mikf committed:
[simpcity] extract attachment/inline files (#8560)

mikf committed:
[simpcity] extract "click to load media" URLs (#8609)

mikf committed:
[twitter] add 'search-results' option (#8613)

mikf committed:
[bunkr] detect when album is deleted mid-download (#8619)

mikf committed:
[output] add 'defer' option for logging files (#8523)

- defer file creation of error files by default
- implement custom FileHandler class
  to support deferred directory creation

mikf committed:
[twitter] restore better user NotFoundError messages (#8621)

fixes regression introduced in 5747dbf00cfad1651b846486a2870b75411f566e

mikf committed:
[twitter] implement 'retries-api' option (#8317)

retry API requests when encountering server-related errors

spicybiguy committed:
[motherless] fix 'gallery_title' extraction (#8605)

* Update motherless.py for title selector

Updated the selector for the title property for Motherless galleries to be an h2 instead of an h1 to reflect changes on the site

* fix 'gallery_title' extraction

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[actions] add 'LoggerAdapter.traceback()' method (#8317)

8c62be343eb8a3b7e3bebdc7566b60c87e4e3666

mikf committed:
[twitter] reset 'api_tries' after successful API call (#8317)

mikf committed:
[pp:exec] add 'verbose' option (#7743)

mikf committed:
[audiochan] add initial support (#8602)

mikf committed:
[schalenetwork] fix 'tags' categories (#8625)

mikf committed:
[civitai] implement extracting 'tags' metadata (#8626)

mikf committed:
[audiochan] add 'search' extractor (#8602)

mikf committed:
[2ch] support 'org' TLD (#8629)

mikf committed:
[twitter] increase 'retries-api' default to '9' (#8317)

mikf committed:
[dl:http] implement '_http_segmented' (#8602)

mikf committed:
[vsco] use '"browser": "firefox"' by default (#8127)

and re-enable TLS 1.2 cipher suites
i.e. revert c7fe6ba691f4bb5c840953b780d31770b6f4cb37

mikf committed:
[discord] limit length of default filenames (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3593460556

mikf committed:
[twitter] update/fix 'pinned' Tweet extraction (#8500)

- handle pins from 'pinned_tweet_ids_str'
- provide 'pinned' metadata

mikf committed:
[twitter] implement using multiple values for 'search-limit' (#8173)

https://github.com/mikf/gallery-dl/issues/8173#issuecomment-3302713173

mikf committed:
[twitter] add 'limit' option (#8173)

- reduce default number of requested items to 50
- allow using "fallback" values like with 'search-limit'

https://github.com/mikf/gallery-dl/issues/8173#issuecomment-3303015837

mikf committed:
[twitter] fix 'AttributeError' for "search-pagination": "max_id" (#8613)

when using it with the regular 'Search' extractor

https://github.com/mikf/gallery-dl/issues/8613#issuecomment-3603946848

mikf committed:
[path] implement conditional 'part-directory' (#8329)

mikf committed:
[fanbox] yield data for inaccessible posts (#8643)

mikf committed:
[dl:ytdl] restructure code

mikf committed:
[dl:ytdl] improve error message when importing default ytdl modules

mikf committed:
[twitter] fix TypeError when '_user_obj' is None (#8648)

fixes regression introduced in 406a74cfaaf9f44003ac2e64583172eaebdb2b0e

mikf committed:
[dl:ytdl] update '_extract_manifest()'

- use 'kwdict' as single argument
- fix '_parse_m3u8_formats_and_subtitles' arguments

mikf committed:
[dl:ytdl] implement 'retry' functionality (#1131 #8269)

for 'info_dict' extraction and file download

mikf committed:
[job] inline 'dispatch' loop

mikf committed:
implement 'post-filter' & 'post-range' options

(#8643)

mikf committed:
[twitter] support '"search-results": "media"' results (#8613 #8635)

https://github.com/mikf/gallery-dl/issues/8635#issuecomment-3615267040

mikf committed:
yield 3-tuples for Message.Directory

adapt tuples to the same length and semantics as other messages

mikf committed:
remove Message.Version

mikf committed:
[oauth] inline/remove top-level imports

mikf committed:
[output] defer creation of 'unsupported' files by default (#8622)

https://github.com/mikf/gallery-dl/pull/8622#issuecomment-3618151200

mikf committed:
[docs/configuration] update 'Logging Configuration' defaults

mikf committed:
[newgrounds] set error status when failed to extract post (#8654)

mikf committed:
[chevereto] extract 'album_id' & 'album_slug' metadata (#8604)

mikf committed:
[tiktok] add 'covers' option (#8515)

icelit4 committed:
[hentaifoundry] extract 'categories' metadata (#8656)

* [hentaifoundry] Add support for categories
* rename 'hf_categories' to 'categories'
* move code into '_extract_categories()' method
* update tests

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[util] support integer values for '…-range' options (#8604)

mikf committed:
[dl:ytdl] fix 'AttributeError' when ytdl returns None (#8660)

fixes regression introduced in 402f53616b149a04a9fed0b5274c4701e78521e7

mikf committed:
[chevereto:album] extract 'album_…' metadata (#8604)

mikf committed:
[common] use 'parent' value as 'parent-metadata' default (#8525 #8604)

enable 'parent-metadata' by default for
'chevereto' & 'imagehost' (album/gallery) extractors

mikf committed:
[twitter] always use '3' as 'search-stop' default (#8613)

https://github.com/mikf/gallery-dl/issues/8613#issuecomment-3626474518

mikf committed:
[twitter] fix "UnboundLocalError: 'last_tweet'" (#8613)

https://github.com/mikf/gallery-dl/issues/8613#issuecomment-3628782844

mikf committed:
[twitter] implement workarounds for empty 'core' data (#8613)

https://github.com/mikf/gallery-dl/issues/8613#issuecomment-3626389269

mikf committed:
[sankaku][idolcomplex] support URLs with locale code (#8667)

mikf committed:
[sankaku][idolcomplex] fix download URLs (#8666)

mikf committed:
[dl:http] fail downloads of empty files (#8661)

mikf committed:
[realbooru] extract video fallbacks (#8661)

mikf committed:
[erome] improve error message for deleted & copyrighted albums (#8665)

mikf committed:
[blueskx] add 'api-server' option (#8668)

mikf committed:
[docs/README] add Docker instructions to pull 'dev' images (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3629069986

mikf committed:
[chevereto:album] extract 'count' & 'num' metadata (#8604)

mikf committed:
[redbust] remove module (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3635792318

mikf committed:
[fikfap] add support (#8673)

add 'post' & 'user' extractors

mikf committed:
[dl:ytdl] forward '_ytdl_manifest_headers' to formats

mikf committed:
[facebook] do not match '/permalink' URLs (#8679)

SpiffyChatterbox committed:
[pornpics] add 'category' and 'listing' extractors (#8662)

* [pornpics] add category and listing extractors

Add support for:
- Category pages like /ass/, /milf/, /blonde/ etc.
- Listing pages like /popular/, /recent/, /rating/, /likes/, /views/, /comments/

Category pages use JSON pagination like tags/search.
Listing pages don't support JSON pagination and use different HTML structure.

* [pornpics] simplify category pattern via class ordering

- Move PornpicsCategoryExtractor after PornpicsListingExtractor
  so it acts as catch-all, eliminating need for negative lookahead
- Use list comprehension in PornpicsListingExtractor.galleries()

* update docs/supportedsites

SpiffyChatterbox committed:
[fitnakedgirls] add support (#8671)

* [fitnakedgirls] add extractor

Add support for fitnakedgirls.com:
- Photo galleries (/photos/gallery/)
- Category pages (/photos/gallery/category/)
- Tag pages (/photos/tag/)
- Video posts (/videos/)
- Blog posts (/fitblog/)

Handles both newer (wp-block-image) and older (size-large) templates.

* simplify & fix

- use '_extract_title' method
- move '_pagination' into base class
- update 'FitnakedgirlsTagExtractor' pattern

* update docs/supportedsites

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[gofile] fix extraction (#8681 #8683)

- fix website_token extraction
- send website_token as 'X-Website-Token' header

mikf committed:
[reddit] guess 'mp4' extension for ytdl downloads (#8684)

SpiffyChatterbox committed:
[nudostarforum] add support (#8664)

Add support for nudostar.com forum (XenForo-based forum site).
This is separate from the existing nudostar.py which handles nudostar.tv.

Supports:
- Thread extraction with pagination
- Individual post extraction
- Authentication via xf_user cookie or username/password
- Internal attachments (both linked and embedded images)
- External image host URLs (queued for recursive processing)

mikf committed:
[xenforo] implement generic XenForo forum extractors

support
- https://simpcity.cr/
- https://nudostar.com/forum/ (#8333)

mikf committed:
[audiochan] extract 'description' texts (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3643689186

mikf committed:
[audiochan] relax 'pattern'

mikf committed:
[dl:ytdl] improve error detection

- intercept ytdl logging messages and signal error when it
  emits an error message
- remove "ERROR:" etc from ytdl logging messages

mikf committed:
[xenforo] emit AuthRequired errors for 403 downloads

mikf committed:
[docs] update 'xenforo' options

mikf committed:
release version 1.31.0

mikf committed:
[audiochan] use proper variable name

mikf committed:
[ytdl] respect '--no-skip'

mikf committed:
[pixiv] warn about invalid 'PHPSESSID' cookie (#8689)

mikf committed:
[2chen] implement generic 2chen board extractors

support
- https://sturdychan.help/
- https://schan.help/ (#8680)

mikf committed:
[misskey] implement 'order-posts' option (#8516)

mikf committed:
[twitter] fix avatar & background downloads with "expand": true (#8698)

mikf committed:
[comedywildlifephoto] add 'gallery' extractor (#8690)

Skaronator committed:
Merge remote-tracking branch 'upstream/master' into ffprobe

mikf committed:
[docs] remove 'twitter.username-alt' entries

mikf committed:
[boosty] warn about expired 'auth' cookie tokens (#8704)

mikf committed:
[koofer] add 'shared' extractor (#8700)

mikf committed:
[mastodon] fix "AttributeError: 'parse_datetime_iso'" (#8709)

fixes regression introduced in c8fc79002808879abfaf1cc307236731169ff427

mikf committed:
[dl:ytdl] fix "UnboundLocalError: 'tries'" (#8707)

fixes regression introduced in 402f53616b149a04a9fed0b5274c4701e78521e7

mikf committed:
[xenforo] support 'allthefallen.moe/forum' (#3249 #8268)

mikf committed:
[boosty] send 'Authorization' header for file downloads (#8704)

mikf committed:
[fanbox] make '_extract_post()' non-fatal (#8711)

fixes regression introduced in 4a4c1a1154af14bf39b74828b9710290029eb78b

mikf committed:
[aryion:favorite] support folder items (#8705)

mikf committed:
[aryion:favorite] support 'category' URLs (#8705)

mikf committed:
[aryion] fix 'description' metadata

mikf committed:
[tenor] fix 'title' metadata

remove the trailing 4 characters only when they are ' GIF'

mikf committed:
[aryion] add 'watch' extractor (#8705)

petercunha committed:
[picazor] add 'user' extractor (#7083 #7504 #7795 #8717)

* [picazor] Add support for Picazor
* [picazor] Improve tests for picazor
* [picazor] run flake8
* [picazor] working on tests
* update extractor
* update tests
* update supportedsites

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[reddit] use '"videos": "dash"' by default (#8657)

https://github.com/mikf/gallery-dl/issues/8657#issuecomment-3663560902

mikf committed:
[formatter] add 'I' format specifier - identity

mikf committed:
[path] combine 'build_*' & 'build_*_conditional'

mikf committed:
[aryion] fix "AttributeError: '_pagination'" (#8723)

fixes regression introduced in 62d6a2206d834315d6c297fdc16ee38aa37c65d4

mikf committed:
[furaffinity] fix 'tags' metadata (#8724)

Co-authored-by: d-koc <141529764+d-koc@users.noreply.github.com>

mikf committed:
[path] implement dynamic length directories (#1350)

append directory segments for each item of a list (or general non-string
iterable), which can be returned with the 'I' specifier

mikf committed:
[aryion:favorite] ignore already seen folders (#8728)

mikf committed:
[tests] add 'path' tests

mikf committed:
[weebdex] add support (#8722)

mikf committed:
[tests:pp] prevent failures due to config dependencies

'filename' & 'directory' set in 'path' tests cause
'ClassifyTest' to fail

mikf committed:
[scripts/release] reduce expected executable sizes

mikf committed:
release version 1.31.1

mikf committed:
[options] add '--sleep-skip' option (#6582)

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3677460485

mikf committed:
Revert "use f-strings when building 'pattern'"

revert d7c97d5a9741470374358d105a6cb0b4a79bff5c.

mikf committed:
use 'operator +' when building 'pattern'

mikf committed:
replace 2-element f-strings with simple '+' concatenations

Python's 'ast' module and its 'NodeVisitor' class
were incredibly helpful in identifying these

mikf committed:
[facebook] add 'loop' option (#8696)

mikf committed:
[subscribestar] add 'user-tag' extractor (#8737)

mikf committed:
[exhentai] fix possible exception in 'finalize()' (#8741)

mikf committed:
[scripts/init] use 'operator +' when building 'pattern'

mikf committed:
Merge branch 'op+': use '+' for 2-element string concatenations

mikf committed:
[subscribestar:user-tag] update 'pattern'

mikf committed:
[imagehosts] remove '_https'

mikf committed:
[imagehosts] remove 'splitext' import

mikf committed:
[discord] add 'server-assets' extractor (#8589)

mikf committed:
[subscribestar] merge 'user-tag' into regular 'user' extractor (#8737)

mikf committed:
[webtoons] download episode background music (#8733)

add 'bgm' option

mikf committed:
[deviantart:stash] fix JSONDecodeEerror for folders (#8750)

remove extra ',"submit":{}' after end of JSON object

mikf committed:
[tests/results] implement more specific '#auth' checks

mikf committed:
[discord] update asset collection & handling (#8589)

mikf committed:
[fansly] disable 'formats' check by default (#8757)

should allow downloading and selecting the highest resolution video format
regardless of its 'type' / format code

mikf committed:
[fansly] add 'previews' option (#8686)

mikf committed:
[actions] fix 'abort' (#8753)

fixes regression introduced in 2eb5e5205590e64813456fc0d63ec7e55794d2f2

forward the correct arguments to StopExtraction.__init__
instead of logging message and 'extras'

mikf committed:
[twitter] add 'notifications' extractor (#7974)

mikf committed:
[instagram:stories] extract correct 'expires' dates (#8764)

mikf committed:
[ytdl] expect all exception types when extracting info_dict (#8343)

not only 'YoutubeDLError's

ollm committed:
[webtoons] extract 'num_play' & 'num_stop' for background music (#8733 #8755)

* [webtoons] play_image and stop_image in background music
* [webtoons] add play_image and stop_image fields to test results
* [webtoons]: flake8
* [webtoons]: flake8 (2)

* update & improve
    - remove global 'images_urls'
    - move enumerate loop into images() extraction
* simplify 'play_image' etc extraction
* rename 'play_image' etc
    - num_play
    - num_stop
    - filename_play
    - filename_stop

---------

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
[imgpv] add 'image' extractor (#8773)

mikf committed:
[dl:ytdl] unset '_mtime_http' (#8767)

prevent an mtime value from a previous HTTP download
from being reused for a file downloaded by 'ytdl'

mikf committed:
[dl:ytdl] support '_ytdl_params'

mikf committed:
[instagram] detect homepage redirects (#8714)

mikf committed:
[tumblr:search] prevent KeyError when using 'offset' pagination (#8720)

mikf committed:
[ahottie] add aupport (#8710)

mikf committed:
[dl:http] add MIME type and signature for .aac files

mikf committed:
[sankaku] fix re-authentication (#8779)

Unset the `Authorization` header before performing a re-login

CasualYT31 committed:
[tiktok] remove yt-dlp dependency & add support for more post types (#8715)

#7246 #8035 #8466 #8730

* [tiktok] support extracting videos directly without yt-dlp
* [tiktok] support extracting users directly without yt-dlp
* [tiktok] fixing logic, tests, linting errors
* [tiktok] implement tiktok-range support for non-yt-dlp user extractor
* [tiktok] Skip range filter if no ranges are given
* [tiktok] Remove debug code
* [tiktok] only check for faulty device IDs during the first couple of passes
    I think the original yt-dlp solution assumes that if a device ID works once, it will always work.
    Plus, my approach would cause needless retries in certain cases if hasMorePrevious does end up being wrong like the original algorithm accounts for. So let's copy the original algorithm here, too.
* [tiktok] support stories
* [tiktok] you can now extract audio without extracting photos
* [tiktok] add TiktokFollowingExtractor
* [tiktok] update supportedsites to include stories
* [tiktok] Keep tiktok-range option for no content user account test
    It acts as a nice guard against that account suddenly having lots of posts to extract
* [tiktok] TiktokUserExtractor and TiktokFollowingExtractor rewrite
* [tiktok] Fix avatar naming convention to match that of posts
* [tiktok] remove type hints for compatibility with older Python versions
* [tiktok] Improve performance of TiktokFollowingExtractor
    This was largely achieved using the story/batch/item_list endpoint
* [tiktok] Forgot to run flake8
* [tiktok] remove old constant
* [tiktok] Support order-posts config item
* [tiktok] flake8
* [tiktok] Older Python versions don't support match
* [tiktok] always ask for posts in chronological order when in "desc" mode
    We should aim to avoid having pinned posts returned before non-pinned ones
* [tiktok] Add liked posts extraction
* [tiktok] Add reposts extraction
* [tiktok] Add saved posts extraction

* cleanup imports
* remove '# MARK:' comments
* remove & simplify 'except' statements
    KeyboardInterrupt & SystemExit inherit from BaseException (not Exception)
    and therefore don't need special handling
* split 'user' extractor
* move PATTERNs into their respective functions
* use dict comprehensions
* add only-matching test URLs for split user extractors
* update config docs
    rename 'tiktok-user-extractor' to 'ytdl'
* document '"popular"' 'order-posts' value
* inline and remove 'util.chunk()'

mikf committed:
[yourlesbians] add 'album' extractor (#8713)

mikf committed:
[tapas] unescape HTML entities in image URLs (#8790)

mikf committed:
[civitai] update 'quality-videos' default (#8787)

from "quality=100" to "original=true,quality=100"
to prevent a server timeout / error

mikf committed:
[xenforo] support '/#post-ID' URLs

mikf committed:
[xenforo] fix incomplete video URLs (#8786)

mikf committed:
[manganelo] add 'bookmark' extractor (#8776)

nthduy committed:
[batoto] replace k-subdomain image URLs with n-subdomain (#8791)

* [batoto] replace k-subdomain image URLs with n-subdomain
    Update tests to use battwo.com proxy domain and current metadata
* [batoto] revert to list comprehension

mikf committed:
[whyp] add support (#8725)

mikf committed:
[weibo] fix 'article' include (#8792)

mikf committed:
[instagram] don't warn on minor size differences (#8300)

only emit a warning with 'warn-images' if the reported original size
is at least 10% smaller then the size of the returned image

mikf committed:
[pixiv] provide 'count' metadata (#8794)

mikf committed:
[pixiv] support 'sketch' include (#8789)

bvergnaud committed:
[bunkr] detect new maintenance video file (#8802)

* fix(bunkr): catch new maintenance video file during validation

Bunkr has performing a large scale maintenance recently, and I've
noticed that I would on occasion download a file only to get
a short server is in maintenance video instead of the intended file.

Debug logs show the filename for the maintenance video has changed:
[urllib3.connectionpool][debug] Starting new HTTPS connection (1): 3d09xl1.b-cdn.net:443
[urllib3.connectionpool][debug] https://3d09xl1.b-cdn.net:443 "GET /c4f36040-bdd1-40b6-aea1-034dfbe88ba2/maint.mp4 HTTP/1.1" 200 322509

This commit fixes the detection of that file to properly send the
"File server in maintenance mode" warning instead of downloading an
unintended file.

Signed-off-by: Benjamin VERGNAUD <ben@bvergnaud.fr>

* check multiple suffixes with one 'endswith()' call

---------

Signed-off-by: Benjamin VERGNAUD <ben@bvergnaud.fr>

mikf committed:
[koofr] implement 'zip' option (#6582 #8700)

Download individual files instead of '.zip' bundle by default.

https://github.com/mikf/gallery-dl/issues/6582#issuecomment-3679504391

mikf committed:
[tiktok] combine '…_id' fields into a single 'file_id' one (#8804)

add a proper 'file_id' value for 'video' files

mikf committed:
[tiktok] update user ID extraction

mikf committed:
release version 1.31.2

Skaronator committed:
Merge tag 'v1.31.2' into ffprobe

version 1.31.2

Skaronator committed:
chore: improve debug logs when ffprobe fails

Skaronator committed:
feat: add ffprobe cache to avoid eccessive reties on download retries
